File: backend/src/forms/forms.controller.ts
import { Controller, Get, Put, Post, Param, Body, UseGuards } from '@nestjs/common';
import { FormsService } from './forms.service';
import { JwtAuthGuard } from '../auth/jwt-auth.guard';

@UseGuards(JwtAuthGuard)
@Controller('forms')
export class FormsController {
  constructor(private formsService: FormsService) {}

  /** GET /forms/template/:nodeId — Get form template for a node */
  @Get('template/:nodeId')
  getTemplate(@Param('nodeId') nodeId: string) {
    return this.formsService.getTemplate(nodeId);
  }

  /** PUT /forms/template/:nodeId — Create/update form template */
  @Put('template/:nodeId')
  upsertTemplate(@Param('nodeId') nodeId: string, @Body() body: { schemaJson: any }) {
    return this.formsService.upsertTemplate(nodeId, body.schemaJson);
  }

  /** POST /forms/submit/:taskId — Submit form data for a task */
  @Post('submit/:taskId')
  submitForm(@Param('taskId') taskId: string, @Body() body: { payloadJson: any; inputMode?: 'MANUAL' | 'VOICE' | 'AI' }) {
    return this.formsService.submitForm(taskId, body.payloadJson, body.inputMode);
  }

  /** GET /forms/submission/:taskId — Get submitted form for a task */
  @Get('submission/:taskId')
  getSubmission(@Param('taskId') taskId: string) {
    return this.formsService.getSubmission(taskId);
  }
}

File: workflow-engine/src/main/java/com/workflow/engine/controller/FormsController.java
package com.workflow.engine.controller;

import com.workflow.engine.model.*;
import com.workflow.engine.repository.CaseRepository;
import com.workflow.engine.repository.PolicyRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

import java.util.Map;
import java.util.UUID;

@RestController
@RequestMapping("/api/forms")
@RequiredArgsConstructor
public class FormsController {

    private final PolicyRepository policyRepository;
    private final CaseRepository caseRepository;

    /** Returns the form template (schemaJson) for a given policy node. */
    @GetMapping("/template/{nodeId}")
    public ResponseEntity<Map<String, Object>> getTemplate(@PathVariable String nodeId) {
        return policyRepository.findByNodeId(nodeId)
                .flatMap(policy -> policy.getNodes().stream()
                        .filter(n -> n.getId().equals(nodeId))
                        .findFirst())
                .filter(node -> node.getFormTemplate() != null)
                .map(node -> ResponseEntity.ok(Map.of("schemaJson", node.getFormTemplate())))
                .orElseGet(() -> ResponseEntity.notFound().build());
    }

    /** Returns the saved form submission for a task. */
    @GetMapping("/submission/{taskId}")
    public ResponseEntity<?> getSubmission(@PathVariable String taskId) {
        return caseRepository.findByTasksId(taskId)
                .flatMap(c -> c.getTasks().stream()
                        .filter(t -> t.getId().equals(taskId))
                        .findFirst())
                .filter(t -> t.getFormSubmission() != null)
                .map(t -> ResponseEntity.ok(t.getFormSubmission()))
                .orElseGet(() -> ResponseEntity.notFound().build());
    }

    /** Saves (or replaces) the form template schema on a policy node. */
    @PutMapping("/template/{nodeId}")
    public ResponseEntity<?> saveTemplate(@PathVariable String nodeId,
                                          @RequestBody Map<String, Object> body) {
        return policyRepository.findByNodeId(nodeId).map(policy -> {
            policy.getNodes().stream()
                    .filter(n -> n.getId().equals(nodeId))
                    .findFirst()
                    .ifPresent(n -> n.setFormTemplate(body.get("schemaJson")));
            policyRepository.save(policy);
            return ResponseEntity.ok(Map.of("ok", true));
        }).orElseGet(() -> ResponseEntity.notFound().build());
    }

    /** Saves (or overwrites) a form submission for a task. */
    @PostMapping("/submit/{taskId}")
    public ResponseEntity<?> submitForm(@PathVariable String taskId,
                                        @RequestBody Map<String, Object> body) {
        return caseRepository.findByTasksId(taskId).map(currentCase -> {
            Task task = currentCase.getTasks().stream()
                    .filter(t -> t.getId().equals(taskId))
                    .findFirst()
                    .orElse(null);
            if (task == null) return ResponseEntity.notFound().build();

            FormSubmission submission = new FormSubmission();
            submission.setId(UUID.randomUUID().toString());
            submission.setPayloadJson(body.get("payloadJson"));

            if (body.containsKey("inputMode")) {
                try {
                    submission.setInputMode(InputMode.valueOf((String) body.get("inputMode")));
                } catch (IllegalArgumentException ignored) {}
            }

            task.setFormSubmission(submission);
            caseRepository.save(currentCase);
            return ResponseEntity.ok((Object) submission);
        }).orElseGet(() -> ResponseEntity.notFound().build());
    }
}
